File: LeetCode/linkedlist/LRUCache.cpp
#include<bits/stdc++.h>

using namespace std;
 struct DListNode {
    int key;
    int value;
    DListNode *next;
    DListNode *prev;
    DListNode() : key(0), value(0), next(nullptr),prev(nullptr) {}
    DListNode(int x,int y) : key(x), value(y), next(nullptr) ,prev(nullptr){}
    DListNode(int x,int y, DListNode *next,DListNode *prev) : key(x), value(y), next(next), prev(prev){}
 };

 struct {
    void print(string msg, DListNode* head){
        DListNode* temp = head;
        cout << msg << endl;
        while(temp!=nullptr){
            cout << temp->key << ": " << temp->value <<endl;
            temp = temp-> next;
        }
        cout << endl;
    }
    void log(string msg){
        cout << "Log:: " << msg <<endl;
    }
    void log(unordered_map<int,DListNode*> hashmap){
        cout << "HASHMAP" << endl;
        for(const auto& it : hashmap){
      
            cout << it.first << ":" << it.second->value << endl;
    }
    }
} logger;


class LRUCache {
public:
    DListNode* head;
    unordered_map<int,DListNode*> hashmap;
    int size;
    LRUCache(int capacity) {
        size = capacity;
        head = nullptr;
         cout << "HASHMAP size " << hashmap.size() <<endl; 
    }
    
    int get(int key) {
        
        if(hashmap[key]){
            SwapNodeFirst(key);
            return hashmap[key]->value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        cout << "HASHMAP size in put " << hashmap.size() <<endl; 
       if(hashmap[key]){
            hashmap[key]->value=value;
            cout << "Calling Swap " << key <<endl;
            SwapNodeFirst(key);
            cout << "Calling Swap??" << endl;   
        }
        else{
            InsertNode(key,value);
        }
    }
    void SwapNodeFirst(int key){
        DListNode* ptr = hashmap[key];
        if(ptr == head) return;
        int value = ptr->value;
        if(ptr->next!=nullptr){
            ptr->next->prev = ptr->prev;
        }
        if(ptr->prev!=nullptr){
            ptr->prev->next = ptr->next;
        }
        ptr->prev = nullptr;
        ptr->next = head;
        if(head!=nullptr){
            head->prev = ptr;
        }
        head = ptr;
        cout << "HASHMAP size before swap " << hashmap.size() <<endl;
        hashmap[key] = head;
        
    }
    void InsertNode(int key,int value){
        logger.print("Insert in ",head);
        if(hashmap.size() > size){
            DeleteLastNode();   
        }
        DListNode *temp = new DListNode(key,value);
        temp->next = head;
        temp->prev = nullptr;
        if(head!=nullptr){
            head->prev = temp;
        }
        head = temp;
        cout << "HASHMAP size before insert " << hashmap.size() <<endl; 
        hashmap[key] = head;
        cout << "HASHMAP size after insert   " << hashmap.size() <<endl;
    }
    void DeleteLastNode(){
        DListNode* curr = head;
        
        while(curr->next!=nullptr && curr->next->next!=nullptr){
            curr= curr->next;
        }if(curr->next!=nullptr){
            cout << " Deleting " << curr->next->key << endl;
            hashmap.erase(curr->next->key);
            curr->next = nullptr;
        }
        else if (curr == head){
            // hashmap.erase(curr->key);
            head = nullptr;
        }
        else{
        }
        logger.log(hashmap);

        
    }
};


int main(){
    int capacity = 1,key=1,value = 1;

    LRUCache* obj = new LRUCache(capacity);
    obj->put(2,1);
    int a = obj->get(2);
    obj->put(3,2);
    a = obj->get(2);
    a= obj->get(3);
    logger.print("Finished",obj->head);
    logger.log(obj->hashmap);
    cout <<"Finished "   <<endl;
    return 0;
}

File: LeetCode/linkedlist/LRUCache.java
import java.util.HashMap;
import java.util.Map;


class LRUCache {
    class DNode{
        int key,value;
        DNode next;
        DNode prev;
        DNode(){}
        DNode(int key,int value){
            this.key = key;
            this.value = value;
            this.next= null;
            this.prev= null;
        }
    }
    DNode head,tail;
    Map<Integer,DNode> cacheMap = new HashMap<Integer,DNode>();
    int size;
    public LRUCache(int capacity) {
        size = capacity;
        head = new DNode();
    }
    
    public int get(int key) {
        if(cacheMap.containsKey(key)){
            SwapNodeToStart(key);
            return cacheMap.get(key).value;
        }
        return -1;
    }
    
    public void put(int key, int value) {
        if(cacheMap.containsKey(key)){
            DNode node = cacheMap.get(key);
            node.value = value;
            SwapNodeToStart(key);
        }
        else{
            InsertNode(key,value);
        }
    }

    public void SwapNodeToStart(int key){
        DNode iNode = cacheMap.get(key);
        if(iNode == head) return;
        if(iNode.next!=null){
            iNode.next.prev = iNode.prev;
        }
        if(iNode.prev!=null){
            iNode.prev.next = iNode.next;
        }
        iNode.prev = null;
        iNode.next = head;
        if(head!=null){
            head.prev = iNode;
        }
        head = iNode;   
        cacheMap.put(key, head);   
    }
    public void InsertNode(int key, int value){
        if(cacheMap.size()== size){
            DeleteNodeFromEnd();
        }
        DNode temp = new DNode(key, value);
        temp.prev = null;
        temp.next = head;
        if(head != null){
            head.prev = temp;
        }
        head = temp;
        cacheMap.put(key, head);
    }
    public void DeleteNodeFromEnd(){
        DNode itr = head;
        display(head);
        while(itr.next!=null && itr.next.next!=null){
            itr = itr.next;
        }
        if(itr.next!=null){
            cacheMap.remove(itr.next.key);
            itr.next = null;
        }
        else if(itr.key == head.key){
            cacheMap.remove(itr.key);
            head = null;
        }
    }

    public static void main(String[] args) {
        int capacity=1;
        LRUCache obj = new LRUCache(capacity);
        
        obj.put(2,1);
        int param_1 = obj.get(2);
        System.out.println("Result Key: 2:"+param_1);
        obj.put(3,2);
        param_1 = obj.get(2);
        System.out.println("Result Key: 2:"+param_1);
        param_1 = obj.get(3);
        System.out.println("Result Key: 3:"+param_1);
        obj.display(obj.head);
    }
    public void display(DNode head){
        DNode temp = head;
        while(temp!=null){
            System.out.println("Node: "+temp.key+":"+temp.value);
            temp = temp.next;
        }
    }
}



